hometask (4)

MihailNenahov opened the pull request:
i tried to write a very cool construct for sorting(i know that in that case we don't realy need to use routines but i tried, to practice it that way).please look what i've done and give some advices, how to do it correctly.

Commit:
finished 4 hometask

wshaman commented on FourthTask/figures/main/main.go:
```diff
@@ -0,0 +1,52 @@
+import (
+	"fmt"
+
+	"github.com/Mishamba/EPAM_GOLANG_COURSE/FourthTask/figures/tests"
```
It's better to keep snake_case in packages in general. e.g. it's OK to use upper case for some abbreviations.

MihailNenahov replied:
I'll consider it in next hometask.

wshaman commented on FourthTask/figures/main/main.go:
```diff
@@ -0,0 +1,52 @@
+)
+
+func main() {
+	if err := tests.Errors(); err != nil { //Programm will start work after tests check
```
it's not the best practice to mix test code and logic.
Usually it solved with CI tools or (sigh...) Makefiles.

wshaman commented on FourthTask/figures/main/main.go:
```diff
@@ -0,0 +1,52 @@
+	tmp, err := c.Area()
+	if err != nil {
+		panic(err)
+	} else {
```
`else` block can be omitted, cuz on panic code will exit. We'd want to keep code as much plain as possible.

wshaman commented on FourthTask/figures/main/main.go:
```diff
@@ -0,0 +1,52 @@
+		panic(err) //If we have errors, we panic. we can't go ahead.
+	}
+
+	tmp, err := c.Area()
```
```suggestion
        var tmp float64
        var err error
	if tmp, err = c.Area(); err != nil {
	    panic(err)
	}
	fmt.Println(tmp)
```
still not the best, just a minor step to reduce amount of code

wshaman commented on FourthTask/figures/tests/tests.go:
```diff
@@ -0,0 +1,51 @@
+func Errors() error {
+	_, err := types.CircleConstruct(-5) //negative input
+	if err != nil {
+		return err
```
Possibly there must be an error. We can't create a circle with negative radius, so this func should return an error, and test should fail if no error was returned, like 
```go
if err := types.CircleConstruct(-5); err == nil {
   panic(errors.New("we are from outer space")
}
```

wshaman commented on FourthTask/figures/tests/tests.go:
```diff
@@ -0,0 +1,51 @@
+//	Tests package. Here i'll write some tests and if smth will go wrong i'll tell main pocket to stop his work
+package tests
```
Refer to lesson 5 on testing please.

wshaman commented on FourthTask/figures/types/types.go:
```diff
@@ -0,0 +1,70 @@
+)
+
+const (
+	circlePanic   string = " circle can't has negative radius"
```
Common practice:
```suggestion
	ErrCircleNegative = errors.New( " circle can't has negative radius")
```
And it will be var, not const
Also, there are some other ways to define errors like this one, feel free to use any you will find suitable.

wshaman commented on FourthTask/figures/types/types.go:
```diff
@@ -0,0 +1,70 @@
+	overflowError string = " result will be bigger, that we can coutn. sorry..."
+)
+
+type Figure interface {
```
Seems you should Export funcs `Area()` and `Perimeter()`  in this interface.
Also, a minor, it's better to keep english names, so `Perimeter` would be better here.

wshaman commented on FourthTask/figures/types/types.go:
```diff
@@ -0,0 +1,70 @@
+	radius int
+}
+
+func CircleConstruct(newRadius int) (Circle, error) {
```
```suggestion
func NewCircle(newRadius int) (Figurine, error) {
```
so, a minor one.  func name could be like `NewStructName` and it must return an interface object `Figurine`

wshaman commented on FourthTask/figures/types/types.go:
```diff
@@ -0,0 +1,70 @@
+
+func CircleConstruct(newRadius int) (Circle, error) {
+	if newRadius < 0 {
+		return Circle{0}, errors.New(circlePanic)
```
```suggestion
		return nil, ErrCircleNegative
```
`nil` can be returned on interfaces and err is already defined, if you'll choose to implement this specific suggestion.

MihailNenahov replied:
i'll correct it, but i don't understand. why it works? `nil` has no `Area` or `Perimetr` functions. or we will return interface with type pointer equal `nil` ?

wshaman replied:
Duh. I don't know where is my answer here.
in short: just ingore it please, I've misread task's description.

wshaman commented on FourthTask/figures/types/types.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+func (c Circle) Area() (float64, error) {
+	if float64(c.radius) > math.Sqrt(9223372036854775807.0/math.Pi) {
```
magic constants are scary. `math.MaxFloat64`

wshaman commented on FourthTask/figures/types/types.go:
```diff
@@ -0,0 +1,70 @@
+	side int
+}
+
+func SquareConstruct(newSide int) (Square, error) {
```
Same as for Circle signature must be used here too.

wshaman commented on FourthTask/sorting/construct/construct.go:
```diff
@@ -0,0 +1,63 @@
+var wg sync.WaitGroup
+
+func People() (res types.People) { //i'll left some questions
+	exit := make(chan string) //this chan i'll use to exit function
```
Seems to be an overkill to use chans/goroutines here.

MihailNenahov replied:
i think, that i woun't to change this file(only if i will have some time). in next task i will use goroutines and it will be really usefull. here i will just make from this goruitines simple functions.

burov replied:
As we are mentioned over engineering it's not a good thing, as well it's hard to as to review a lot of creative ideas where people just want to bring some technology to task just for fun

imatsiash-ib replied:
Please re-implement it without go-routines.

wshaman commented on FourthTask/sorting/tests/tests.go:
```diff
@@ -0,0 +1,38 @@
+	"github.com/Mishamba/EPAM_GOLANG_COURSE/FourthTask/sorting/types"
+)
+
+func Errors() error {
```
Same about tests, please refer lesson 5

wshaman commented on FourthTask/sorting/construct/construct.go:
```diff
@@ -0,0 +1,63 @@
+			var newYear, newMonth, newDay int
+
+			fmt.Println("Enter Person")
+			fmt.Print("First and Last name: ") //here. i don't know why it happens(i can't deug my code, don't know, how to create config for goland)
```
Just switch package to main and add a main func, the simplest solution is (usually) the best. :wink:

MihailNenahov replied:
but what about rule, that better to create for each package new folder?

burov replied:
@Mishamba it's depends from scope of task if it's a really simple program less than 400 LOC it's preferable to use one file

wshaman commented on FourthTask/sorting/construct/construct.go:
```diff
@@ -0,0 +1,63 @@
+			}
+
+			fmt.Print("Birtday year, month, day: ")
+			_, err = fmt.Scan(&newYear, &newMonth, &newDay)
```
```suggestion
			if _, err = fmt.Scan(&newYear, &newMonth, &newDay); err != nil {
			    panic(err)
			}
```
Again, just a minor code readability improvement.

wshaman commented on FourthTask/sorting/main/main.go:
```diff
@@ -0,0 +1,34 @@
+package main
```
You don't have to create a folder `main` to hold a package `main`.
Feel free to create a separate folder for a task, create a `main` package there and subpackages with the names you want. 
You can ping me in slack if you need some help on it.

wshaman commented on FourthTask/sorting/types/types.go:
```diff
@@ -0,0 +1,40 @@
+)
+
+var (
+	StartDate time.Time = time.Date(0, time.January, 1, 0, 0, 0, 0, time.UTC) //user will set data relatively Jesus birthday
```
There are more important dates in go world, like "2006-01-02 15:04:05" for example. 
But usually you want to set auto zero-nil value for unset variables. 
like 
`var StartDate time.Time` is OK most of the times you need it. Any value set to some non-default value could lead to hardly debugable errors.

wshaman commented on FourthTask/sorting/types/types.go:
```diff
@@ -0,0 +1,40 @@
+
+func (p People) Less(i, j int) bool {
+
+	diffDays := p[j].Birthday.Sub(p[i].Birthday) / 24
```
I bet this should work, but you can use built-in magic https://golang.org/pkg/time/#Time.Before

MihailNenahov replied:
in this task i can't use this func. because we have 3 cases here.

wshaman replied:
1st check if equals, 2nd if Before, for example. Anyway, it's up to you which way to prefer.

wshaman reviewed (changes requested):
good start!
Please, take a look at testing.
4.1 should be simplified, I think. but it's OK if tests will pass.
Feel free to ask in slack, 
upd: I've removed my comments on figurines creating. Mea culpa.

wshaman commented:
It was my misread of initial task, sorry.
Correct, struct can't be nil w/o a pointer, this comment is valid for
interface only.
Feel free to ignore it.

чт, 28 нояб. 2019 г., 0:14 Misha Nenahov <notifications@github.com>:

> *@Mishamba* commented on this pull request.
> ------------------------------
>
> In FourthTask/figures/types/types.go
> <https://github.com/Mishamba/EPAM_GOLANG_COURSE/pull/8#discussion_r351500171>
> :
>
> > +	squarePanic   string = " square can't has negative side"
> +	overflowError string = " result will be bigger, that we can coutn. sorry..."
> +)
> +
> +type Figure interface {
> +	area() (float64, error)
> +	perimetr() (float64, error)
> +}
> +
> +type Circle struct {
> +	radius int
> +}
> +
> +func CircleConstruct(newRadius int) (Circle, error) {
> +	if newRadius < 0 {
> +		return Circle{0}, errors.New(circlePanic)
>
> i'll correct it, but i don't understand. why it works? nil has no Area or
> Perimetr functions. or we will return interface with type pointer equal
> nil ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mishamba/EPAM_GOLANG_COURSE/pull/8?email_source=notifications&email_token=AAA36AUMNQMFRAJP6ROCQGLQV3PK5A5CNFSM4JRNFPKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCNHQ3GI#discussion_r351500171>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA36AR7VDBXBWKRHYPDIPDQV3PK5ANCNFSM4JRNFPKA>
> .
>

Commit:
corrected everythink except tests

Mishamba commented:
because of my inadvertence i added in last commit some files from next hometask.

Commit:
finished tests

burov commented on FourthTask/sorting/types/types.go:
```diff
@@ -0,0 +1,46 @@
+
+func (p People) Less(i, j int) bool {
+	result := p[i].Birthday.Before(p[j].Birthday)
+	return result
```
Less function implemented not properly

burov reviewed (commented):
Please see comments for task about sort.Interface, figures, and print smile ok for me

Commit:
changed Less func

wshaman commented on FifthTask/debugSomeCode/someCodeToDebug.go:
```diff
@@ -0,0 +1,47 @@
+package main
```
`go vet` shows some warnings `go build` ignores. Use `go vet` please.

MihailNenahov replied:
smth wrong. i used `go vet` but nothing happened.

wshaman reviewed (commented):
I'd say 4th is good to go, but can't approve because of mixing HW 4/5.

Mishamba commented:
> I'd say 4th is good to go, but can't approve because of mixing HW 4/5.

Commit:
removed fifth hw

Mishamba commented:
> Please see comments for task about sort.Interface, figures, and print smile ok for me

Hello World 😄 !

burov commented on FourthTask/figures/types/types.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+func (c Circle) Area() (float64, error) {
+	return math.Pi * math.Pow(float64(c.radius), 2), nil
```
It's seems for me it's not required there, it's preferable to use one of the approach(validate data on initialisation or validate on computation) both seems overkill

Mishamba commented:
> I'd say 4th is good to go, but can't approve because of mixing HW 4/5.

if i still can to improve my task, show how to do.

Mishamba commented:
> > I'd say 4th is good to go, but can't approve because of mixing HW 4/5.

if i still can to improve my task, show what to do.

Commit:
going to rewrite it another way

Commit:
5 - 1 task v 1.0

Commit:
finished rewriting and wrote tests(92.6% cover)

Commit:
finished 4 HW construct

Commit:
going to change my code for better cover

wshaman commented on FourthTask/sorting/types/types_test.go:
```diff
@@ -0,0 +1,52 @@
+	sort.Sort(somePersons)
+	//sortPersons = NewPeople(, person4) //need to write it
+	key = equalSlice(somePersons, somePersons)
+	assert.True(t, key, "sort work's wrong way 3")
```
`assert.Equal` can compare slices. But that's right, usually it's better to make a simple comparing solution.

imatsiash-ib commented on FourthTask/figures/types/coverout.html:
```diff
@@ -0,0 +1,172 @@
+
```
Not sure that these files should be uploaded to the github.

imatsiash-ib commented on FourthTask/figures/types/types.go:
```diff
@@ -0,0 +1,58 @@
+
+type Figure interface {
+	Area() (float64, error)
+	Perimetr() (float64, error)
```
Do we really need `error` result in these functions ?  You always return `nil` as a second result value.

imatsiash-ib commented on FourthTask/figures/types/types_test.go:
```diff
@@ -0,0 +1,74 @@
+	}
+	perimetr, err = square.Area()
+	assert.Equal(t, perimetr, 0, "Square perimetr exeption - overflow check not works")
+	assert.NotNil(t, err, "Square perimetr exeption - overflow check not works")*/
```
All useless comments should be removed from PR.

imatsiash-ib commented on FourthTask/sorting/types/types.go:
```diff
@@ -0,0 +1,59 @@
+	"time"
+)
+
+var StartDate time.Time = time.Date(0, time.January, 1, 0, 0, 0, 0, time.UTC) //user will set data relatively Jesus birthday
```
Not sure that we need define this time explicitly.

imatsiash-ib commented on FourthTask/sorting/types/types.go:
```diff
@@ -0,0 +1,59 @@
+	var tmp Person
+	tmp = p[i]
+	p[i] = p[j]
+	p[j] = tmp
```
you can use `p[i], p[j] = p[j], p[i]`

imatsiash-ib commented on FourthTask/sorting/types/types.go:
```diff
@@ -0,0 +1,59 @@
+	p[j] = tmp
+}
+
+func NewPeople(args ...Person) (result People) { //this fucn was created for tests
```
If this function is used only for tests - will be better to define it inside the test package.

imatsiash-ib commented on FourthTask/sorting/types/types.go:
```diff
@@ -0,0 +1,59 @@
+	}
+
+	return result
+}
```
I don't think that this function makes much sense. Will be better to define slice explicitly.

Commit:
after reading comments

Commit:
deleted redundant files

imatsiash-ib commented on FourthTask/figures/types/types.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+type Figure interface {
+	Area() (float64, error)
```
Your objects do not implement `Figure` interface. Interface methods return 2 values, your methods return 1 value.

Commit:
after reading comments

burov commented on FourthTask/sorting/construct/construct.go:
```diff
@@ -0,0 +1,62 @@
+//it works, but i have one problem there
```
Still don't understand what the reason for this file

MihailNenahov replied:
it's just a construct with user interface. i thought that it's cool idea to work with goroutines in this file.

Commit:
after reading comments

removed repository with construct
tried to remove unnecessary error check, but when i cloned branch i found out that i already removed it. if u can't see it, please write it as a comment
